experimental/oauther/oauth2.go

```diff
@@ -0,0 +1,173 @@
+	// DefaultOAuth2StateTimeToLive is the duration the states from the OAuth flow will live in the KVStore by default.
+	DefaultOAuth2StateTimeToLive = 5 * time.Minute
+)
+
+const (
+	connectURL  = "/connect"
+	completeURL = "/complete"
+)
+
+// OAuther defines an object able to perform the OAuth flow.
+type OAuther interface {
+	// GetToken returns the oauth token for userID, or error if it does not exist or there is any store error.
+	GetToken(userID string) (*oauth2.Token, error)
+	// GetConnectURL returns the URL to reach in order to start the OAuth flow.
+	GetConnectURL() string
+	// Deauthorize removes the token for userID. Return error if there is any store error.
+	Deauthorize(userID string) error
+	// ServeHTTP implements http.Handler
+	ServeHTTP(w http.ResponseWriter, r *http.Request)
+}
+
+type oAuther struct {
+	pluginURL             string
+	config                oauth2.Config
+	onConnect             func(userID string, token *oauth2.Token)
+	store                 common.KVStore
+	logger                logger.Logger
+	storePrefix           string
+	oAuthURL              string
+	connectedString       string
+	oAuth2StateTimeToLive time.Duration
+}
+
+/*
+New creates a new OAuther.
+
+- pluginURL: The base URL for the plugin (e.g. https://www.instance.com/plugins/pluginid).
+
+- oAuthConfig: The configuration of the Authorization flow to perform.
+
+- onConnect: What to do when the Authorization process is complete.
+
+- store: A KVStore to store the data of the OAuther.
+
+- l Logger: A logger to log errors during authorization.
+
+- options: Optional options for the OAuther. Available options are StorePrefix, OAuthURL, ConnectedString and OAuth2StateTimeToLive.
+*/
+func New(
+	pluginURL string,
+	oAuthConfig oauth2.Config,
+	onConnect func(userID string, token *oauth2.Token),
+	store common.KVStore,
+	l logger.Logger,
+	options ...Option,
+) OAuther {
+	o := &oAuther{
+		pluginURL:             pluginURL,
+		config:                oAuthConfig,
+		onConnect:             onConnect,
+		store:                 store,
+		logger:                l,
+		storePrefix:           DefaultStorePrefix,
+		oAuthURL:              DefaultOAuthURL,
+		connectedString:       DefaultConnectedString,
+		oAuth2StateTimeToLive: DefaultOAuth2StateTimeToLive,
+	}
+
+	for _, option := range options {
+		option(o)
+	}
+
+	o.config.RedirectURL = o.pluginURL + o.oAuthURL + "/complete"
+
+	return o
+}
+
+/*
+NewFromClient creates a new OAuther from the plugin api client.
+
+- pluginapi: A plugin api client.
+
+- pluginID: The plugin ID.
+
+- oAuthConfig: The configuration of the Authorization flow to perform.
+
+- onConnect: What to do when the Authorization process is complete.
+
+- l Logger: A logger to log errors during authorization.
+
+- options: Optional options for the OAuther. Available options are StorePrefix, OAuthURL, ConnectedString and OAuth2StateTimeToLive.
+*/
+func NewFromClient(
+	client *pluginapi.Client,
+	pluginID string,
```

Comment:
Actually, there is a `GetManifest` method in `pluginapi.Client`. Hence `pluginID` isn't needed.

Reply:
Not sure how I did not find it. Thanks!